SignalServiceKit/src/Storage/KeychainStorage.swift

```diff
@@ -0,0 +1,103 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import SAMKeychain
+
+public enum KeychainStorageError: Error {
+    case failure(description: String)
+}
+
+// MARK: -
+
+@objc public protocol KeychainStorage: class {
+
+    @objc func string(forKey key: String, service: String) throws -> String
+
+    @objc func set(string: String, forKey key: String, service: String) throws
+
+    @objc func data(forKey key: String, service: String) throws -> Data
+
+    @objc func set(data: Data, forKey key: String, service: String) throws
+
+    @objc func remove(key: String, service: String) throws
+}
+
+// MARK: -
+
+@objc
+public class SSKKeychainStorage: NSObject, KeychainStorage {
+
+    @objc public static let sharedInstance = SSKKeychainStorage()
+
+    // Force usage as a singleton
+    override private init() {
+        super.init()
+
+        SwiftSingletons.register(self)
+    }
+
+    @objc public func string(forKey key: String, service: String) throws -> String {
+        var error: NSError?
+        let result = SAMKeychain.password(forService: service, account: key, error: &error)
+        if let error = error {
+            throw SSKProtoError.invalidProtobuf(description: "\(logTag) error retrieving string: \(error)")
```

Comment:
I don't think these are supposed to be SSKProtoError's, are they?

TIOLI: I often make a private factory method like:

```
private func keychainError(_ description: String) -> KeyChainError {
    return KeyChainError.failure(description: description)
}
```

Reply:
> I don't think these are supposed to be SSKProtoError's, are they?

Done.